src/pytest_inmanta_lsm/remote_service_instance_async.py

```diff
@@ -0,0 +1,544 @@
+
+        # Save the start time to know when we should trigger a timeout error
+        start = time.time()
+
+        # Save the last state, for logging purpose, to tell the user every time we meet a new state
+        last_state: typing.Optional[str] = None
+
+        # Save the last version we treated, to avoid going through the full history at every
+        # iteration
+        last_version = start_version
+        while True:
+            # Go through each log since the last iteration, starting from the oldest
+            # states, including the last version we controlled at the previous iteration
+            # to make sure the list returned by the server is not empty
+            # cf. https://github.com/inmanta/inmanta-lsm/issues/1635
+            for log in sorted(
+                await self.history(since_version=last_version),
+                key=lambda log: log.version,
+            ):
+                try:
+                    # Always skip the last version, as it is either our start version, or a
+                    # version we checked on the previous iteration.
+                    if log.version > last_version and is_done(log):
+                        return await self.get()
+                except BadStateError:
+                    # We encountered a bad state, print the diagnosis then quit
+                    diagnosis = await self.remote_orchestrator.request(
+                        "lsm_services_diagnose",
+                        FullDiagnosis,
+                        tid=self.remote_orchestrator.environment,
+                        service_entity=self.service_entity_name,
+                        service_id=self.instance_id,
+                        version=log.version,
+                        rejection_lookbehind=self._lookback - 1,
+                        failure_lookbehind=self._lookback,
+                    )
+                    LOGGER.info(
+                        "Service instance %s reached bad state %s: \n%s",
+                        self.instance_name,
+                        log.state,
+                        devtools.debug.format(diagnosis),
+                    )
+                    raise
+
+                if last_state != log.state:
+                    # We reached a new state, log it for the user
+                    LOGGER.debug(
+                        "Service instance %s moved to state %s (version %s)",
+                        self.instance_name,
+                        log.state,
+                        log.version,
+                    )
+                    last_state = log.state
+
+                # Save the current version
+                last_version = log.version
+
+            if time.time() - start > timeout:
+                # We reached the timeout, we should stop waiting and raise an exception
+                LOGGER.info(
+                    "Service instance %s exceeded timeout while waiting for %s, current state is %s",
+                    self._instance_name,
```

Comment:
This could be none if we call this method before creating something
But I guess you wanted to do something like this?
```suggestion
                    self.instance_name,
```

Reply:
Fixed